core/src/multi_proof.rs

```diff
@@ -0,0 +1,548 @@
+//! Generate a multiproof from a vector of path proofs.
+//! The multiproof will contain the minimum information needed to verify
+//! the inclusion of all provided path proofs.
+
+use crate::{
+    proof::{PathProof, PathProofTerminal},
+    trie::Node,
+};
+use bitvec::{order::Msb0, view::BitView};
+
+// Each terminal node may have a set of siblings that are uniquely
+// used during the verification with its terminal node.
+// This struct contains the terminal node itself,
+// the depth at which the siblings start to be uniquely associated to that terminal,
+// and the siblings themselves.
+#[derive(Debug, Clone)]
+struct SubPathProof {
+    terminal: PathProofTerminal,
+    depth: u8,
+    inner_siblings: Vec<Node>,
+}
+
+/// A proof of multiple path through the trie.
+#[derive(Debug, Clone)]
+pub struct MultiProof {
+    // All subpaths related to a single terminal node
+    sub_paths: Vec<SubPathProof>,
+    // Vector containing the minimum number of nodes required
+    // to reconstruct all other nodes later.
+    external_siblings: Vec<Node>,
+}
+
+impl MultiProof {
+    /// Construct a MultiProof from a vector of *ordered* PathProof
+    pub fn from_path_proofs(path_proofs: Vec<PathProof>) -> Self {
```

Comment:
this function would be more readable if it were split into multiple helper functions.